[hipblasLT] SK use Tree reduction if < 2 split

daineAMD opened the pull request:
## Motivation

If origami gives us Parallel reduction and sk.grid/tiles==1 then we will launch post kernel call, this PR sets reduction strategy to Tree reduction in this case.

## Technical Details

After calling getSKReduction, this does a check to see if sk.grid/tiles < 2, if so we ensure the reduction type is Tree reduction. Small change to origami intended to keep `select_grid` behaviour the same.

## Test Plan

hipblaslt-test ran locally and passed, wait for CI, need to verify downstream issue is resolved.

## Test Result

hipblaslt-test passed, awaiting other results.

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.
ROCM-8668

daineAMD committed:
[hipblasLT] SK use Tree reduction if < 2 split

daineAMD committed:
Revert "[hipblasLT] SK use Tree reduction if < 2 split"

This reverts commit ba3e37c6a9b25759b22cab63c77499e705d3ced6.

daineAMD committed:
[origami] use cu_count when determining reduction method to avoid parallel reduction with < 4 splits

bstefanuk commented:
Question: How is this related to CPX mode?

bstefanuk reviewed (approved):
The description and title refer to "tree reduction", but it seems like it's an error print, an assert, and a divide. Is it the `cu_count / 4` statement that's technically the tree reduction? Bear with me, just trying to learn some terms here.

nakajee commented on projects/hipblaslt/tensilelite/src/ContractionSolution.cpp:
```diff
@@ -1994,6 +1994,12 @@ namespace TensileLite
         }
         rv.kernelName = outputConversionKernelName(problem, inputs, vw, gsu);
 
+        if(sizeMapping.streamK > 0 && gsu < 2)
```
I found the same gsu calculation for streamK in outputConversionCallArgs (called from here and grouped GEMM case).
I think it is inconsistent if we do this error check only this location.

nakajee commented on projects/hipblaslt/tensilelite/src/ContractionSolution.cpp:
```diff
@@ -1994,6 +1994,12 @@ namespace TensileLite
+        if(sizeMapping.streamK > 0 && gsu < 2)
+        {
+            std::cerr << "\nhipblasLT Error: Trying to launch kernel " << rv.kernelName << " with StreamK and gsu < 2\n";
+            assert(!(sizeMapping.streamK > 0 && gsu < 2));
```
I do not see cerr print + assert in this file.
Mostly, using throw std::runtime_error.

Any reason we use cerr + assert here?

Can't we continue to run by adjusting gsu value here?

daineAMD replied:
Thanks for the quick review Koji.
I'll change to `runtime_error`.
The only way we get here with streamK && gsu < 2 is if sk.reduction == Parallel (https://github.com/ROCm/rocm-libraries/blob/develop/projects/hipblaslt/tensilelite/src/ContractionSolution.cpp#L2618), which should never be the case. Maybe it makes more sense to do this earlier in `solve()`. I'll make a change quick and request a re-review.

amd-jnovotny commented:
@daineAMD Does this need a changelog entry? Also, if this is for 7.2.1, the hipBLASLt version should be bumped somewhere (there are a couple of other PRs for the component for 7.2.1, but it doesn't look it's bumped anywhere.

daineAMD committed:
Change to runtime_error and move check right after origami call

daineAMD committed:
Update versions for ROCm 7.2.1 and update changelog

daineAMD commented:
> @daineAMD Does this need a changelog entry? Also, if this is for 7.2.1, the hipBLASLt version should be bumped somewhere (there are a couple of other PRs for the component for 7.2.1, but it doesn't look it's bumped anywhere.

I've added version bumps for hipblasLT and Origami. I updated the changelog for hipblasLT, I don't see a changelog for origami directly but I think the doc added here should be sufficient.

amd-jnovotny commented on projects/hipblaslt/CHANGELOG.md:
```diff
@@ -2,6 +2,12 @@
+
+### Resolved issues
+
+* Fix issue where users may encounter HIPBLAS_STATUS_INTERNAL_ERROR with various sizes in CPX mode
```
```suggestion
* Fix issue where users might encounter a `HIPBLAS_STATUS_INTERNAL_ERROR` with various sizes in CPX mode.
```

amd-jnovotny reviewed (commented):
One minor for the CHANGELOG. Good otherwise.

daineAMD committed:
Update projects/hipblaslt/CHANGELOG.md

Co-authored-by: Jeffrey Novotny <jnovotny@amd.com>

nakajee reviewed (approved):
Thanks for your update.
Looks good.
